### PR TITLE
added hackpad keyboard

### PR DESCRIPTION
**Pull request by ThePurox**
## Description
I've designed a 4x4 numpad/macropad with an rotary encoder and per key RGB-LEDs. 
It is designed for a workshop at the local Arduino Day in Bayreuth Germany about mechanical keyboards. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/hackpad/config.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#define RGBLIGHT_EFFECT_CHRISTMAS
+#define RGBLIGHT_EFFECT_STATIC_GRADIENT
+#define RGBLIGHT_EFFECT_RGB_TEST
+#define RGBLIGHT_EFFECT_ALTERNATING
+#define RGBLIGHT_EFFECT_TWINKLE
+#define RGBLIGHT_LIMIT_VAL 255
+#define RGB_VAL_STEP 12
+#define RGBLIGHT_DEFAULT_SPD 144
+#define RGBLIGHT_DEFAULT_MODE RGBLIGHT_MODE_RAINBOW_SWIRL + 2
+#endif
```
Comment (elpekenin) [inline keyboards/hackpad/config.h:42]: Most (if not all) of this can be moved to json

File: keyboards/hackpad/rules.mk [hunk 2]
```diff
@@ -0,0 +1,2 @@
+RGBLIGHT_ENABLE = yes
+ENCODER_ENABLE = yes
```
Comment (elpekenin) [inline keyboards/hackpad/rules.mk:2]: JSON too

**Review by elpekenin — CHANGES_REQUESTED** [review]

File: keyboards/hackpad/config.h (hunk 1 above)
Comment (ThePurox) [reply keyboards/hackpad/config.h]: Now it is

File: keyboards/hackpad/rules.mk (hunk 2 above)
Comment (ThePurox) [reply keyboards/hackpad/rules.mk]: Now it is

**Comment by ThePurox** [comment]
It builds fine on my machine. Does someone know how to fix the CI Build failure? `pin` is defined in `info.json`, but the CI complaines that `WS2812_DI_PIN` is undeclared.

File: keyboards/hackpad/info.json [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    "processor": "atmega32u4",
+    "url": "https://git.imaginaerraum.de/Purox/Hackpad",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x0000",
+        "vid": "0xFEED"
+    },
+    "rgblight" : {
+        "led_count" : 16,
+        "pin": "D4",
```
Comment (sigprof) [inline keyboards/hackpad/info.json:35]:
```suggestion
    "ws2812": {
        "pin": "D4"
    },
    "rgblight" : {
        "led_count" : 16,
```
Your PR targets the `develop` branch, where `rgblight.pin` has been replaced by `ws2812.pin` (#20303).

File: keyboards/hackpad/info.json [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            "knight"  : true,
+            "christmas" : true,
+            "static_gradient" : true,
+            "rgb_test" : true,
+            "alternating" : true,
+            "twinkle" : true
+        },
+    },
+    "layouts": {
+        "LAYOUT_numpad_4x4": {
```
Comment (sigprof) [inline keyboards/hackpad/info.json:55]:
```suggestion
        "LAYOUT": {
```
`LAYOUT_numpad_4x4` looks [like this](https://github.com/qmk/qmk_firmware/blob/d3b8179f467da23fec98e97b28684c43e9e8ba73/layouts/default/numpad_4x4/default_numpad_4x4/keymap.c), which does not match your board.  There does not seem to be any well-known layout that matches your board, so the layout macro name should be just `LAYOUT`; the rest of code needs to be changed accordingly at the same time.

File: keyboards/hackpad/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+     * │ 7 │ 8 │ 9 │ + │
+     * ├───┼───┼───┼───┤
+     * │ 4 │ 5 │ 6 │ - │
+     * ├───┼───┼───┼───┤
+     * │ 1 │ 2 │ 3 │ * │
+     * ├───┼───┼───┼───┤
+     * │ 0 │ . │Ent│ / │
+     * └───┴───┴───┴───┘
+     */
+    [0] = LAYOUT_numpad_4x4(                      LT(1, KC_MUTE),
```
Comment (sigprof) [inline keyboards/hackpad/keymaps/default/keymap.c:20]:
```suggestion
    [0] = LAYOUT(                                 LT(1, KC_MUTE),
```

File: keyboards/hackpad/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                            KC_7, KC_8,   KC_9,   DE_PLUS,
+                            KC_4, KC_5,   KC_6,   DE_MINS,
+                            KC_1, KC_2,   KC_3,   DE_ASTR,
+                            KC_0, KC_DOT, KC_ENT, DE_SLSH),
+
+    [1] = LAYOUT_numpad_4x4(                              XXXXXXX,
```
Comment (sigprof) [inline keyboards/hackpad/keymaps/default/keymap.c:26]:
```suggestion
    [1] = LAYOUT(                                         XXXXXXX,
```

File: keyboards/hackpad/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+                            RGB_TOG,  RGB_SAI,  RGB_SAD,  RGB_VAD)};
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (clockwise) {
+        tap_code_delay(KC_VOLU, 10);
+    } else {
+        tap_code_delay(KC_VOLD, 10);
+    }
+    return false;
+}
```
Comment (sigprof) [inline keyboards/hackpad/keymaps/default/keymap.c:39]: The default encoder behavior needs to be implemented at the keyboard level (implement `encoder_update_kb()` in `keyboards/hackpad/hackpad.c`); this is required to make encoders do something in configurator builds (the configurator does not have encoder support at the moment, so it relies on `encoder_update_kb()` being implemented at the keyboard level).

However, you may consider using `ENCODER_MAP_ENABLE = yes` in keymaps (but enabling that option at the keyboard level is not allowed).  This is also the only allowed way to support encoders in VIA, if you choose to add VIA support to your board.

File: keyboards/hackpad/config.h
Comment (sigprof) [inline keyboards/hackpad/config.h:1]: If you don't need to add any defines to `config.h`, remove that file completely (unlike `rules.mk`, the presence of `config.h` is not required).

File: keyboards/hackpad/info.json [hunk 8]
```diff
@@ -0,0 +1,77 @@
+{
+    "manufacturer": "Nico Stuhlmueller",
+    "keyboard_name": "hackpad",
+    "maintainer": "ThePurox",
+    "bootloader": "caterina",
```
Comment (sigprof) [inline keyboards/hackpad/info.json:5]:
```suggestion
    "development_board": "promicro",
```
Setting `"development_board": "promicro"` automatically sets `processor`, `bootloader` and `pin_compatible` appropriately; the `pin_compatible` setting is needed to make [converters](https://docs.qmk.fm/#/feature_converters) work with your board.

File: keyboards/hackpad/info.json [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    "matrix_pins": {
+        "cols": ["B1", "B3", "B2", "B6"],
+        "rows": ["C6", "D7", "E6", "B4", "B5"]
+    },
+    "encoder" : {
+        "rotary" : [
+            {"pin_a" : "F5", "pin_b" : "F4"}
+        ]
+    },
+    "processor": "atmega32u4",
```
Comment (sigprof) [inline keyboards/hackpad/info.json:26]:
```suggestion
```
No longer needed with `"development_board": "promicro"`.

File: keyboards/hackpad/readme.md [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+    make hackpad:default:flash
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
+
+## Bootloader
+
+Enter the bootloader in 2 ways:
+
+* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
```
Comment (sigprof) [inline keyboards/hackpad/readme.md:30]: Looks like your board has a reset button on the top side, so this part of the template needs changes.

**Review by sigprof — CHANGES_REQUESTED** [review]

File: keyboards/hackpad/keymaps/default/keymap.c (hunk 7 above)
Comment (ThePurox) [reply keyboards/hackpad/keymaps/default/keymap.c]: moved it

File: keyboards/hackpad/config.h
Comment (ThePurox) [reply keyboards/hackpad/config.h:1]: done

File: keyboards/hackpad/readme.md (hunk 10 above)
Comment (ThePurox) [reply keyboards/hackpad/readme.md]: Wow, you looked really closely! Thank you for that! I've changed it accordingly.

File: keyboards/hackpad/hackpad.c [hunk 11]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2023 Nico Stuhlmueller (@ThePurox)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+
```
Comment (waffle87) [inline keyboards/hackpad/hackpad.c:5]: This is the default/core behaviour on the develop branch, so this file can be removed.

File: keyboards/hackpad/info.json [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    },
+    "ws2812": {
+        "pin": "D4"
+    },
+    "rgblight" : {
+        "led_count" : 16,
+        "hue_steps" : 20,
+        "brightness_steps" : 10,
+        "saturation_steps" : 20,
+        "max_brightness" : 255,
```
Comment (zvecr) [inline keyboards/hackpad/info.json:40]:
```suggestion
```

File: keyboards/hackpad/info.json [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        "vid": "0xFEED"
+    },
+    "ws2812": {
+        "pin": "D4"
+    },
+    "rgblight" : {
+        "led_count" : 16,
+        "hue_steps" : 20,
+        "brightness_steps" : 10,
+        "saturation_steps" : 20,
```
Comment (zvecr) [inline keyboards/hackpad/info.json:39]: These values differ so little from the defaults that they dont add a ton of value
```suggestion
```

File: keyboards/hackpad/info.json [hunk 14]
```diff
@@ -0,0 +1,78 @@
+{
+    "manufacturer": "Nico Stuhlmueller",
+    "keyboard_name": "hackpad",
+    "maintainer": "ThePurox",
+    "development_board": "promicro",
+    "diode_direction": "COL2ROW",
+    "features": {
+        "bootmagic": false,
```
Comment (zvecr) [inline keyboards/hackpad/info.json:8]:
```suggestion
        "bootmagic": true,
```

File: keyboards/hackpad/info.json [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": false,
```
Comment (zvecr) [inline keyboards/hackpad/info.json:12]:
```suggestion
        "mousekey": true,
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/hackpad/hackpad.c (hunk 11 above)
Comment (ThePurox) [reply keyboards/hackpad/hackpad.c]: Thank you. I've deleted it.

File: keyboards/hackpad/info.json (hunk 12 above)
Comment (ThePurox) [reply keyboards/hackpad/info.json]: done

File: keyboards/hackpad/info.json (hunk 13 above)
Comment (ThePurox) [reply keyboards/hackpad/info.json]: done

File: keyboards/hackpad/info.json (hunk 15 above)
Comment (ThePurox) [reply keyboards/hackpad/info.json]: done

**Comment by ThePurox** [comment]
Is there anything left to do to get this PR merged?

**Review by waffle87 — APPROVED** [review]

File: keyboards/hackpad/readme.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
+# hackpad
+
+![hackpad](https://git.imaginaerraum.de/Purox/Hackpad/src/branch/main/pic.jpg)
```
Comment (fauxpark) [inline keyboards/hackpad/readme.md:3]: Dead link.

File: keyboards/hackpad/info.json [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+                { "matrix": [2, 2], "x": 2, "y": 2 },
+                { "matrix": [2, 3], "x": 3, "y": 2 },
+                { "matrix": [3, 0], "x": 0, "y": 3 },
+                { "matrix": [3, 1], "x": 1, "y": 3 },
+                { "matrix": [3, 2], "x": 2, "y": 3 },
+                { "matrix": [3, 3], "x": 3, "y": 3 },
+                { "matrix": [4, 0], "x": 0, "y": 4 },
+                { "matrix": [4, 1], "x": 1, "y": 4 },
+                { "matrix": [4, 2], "x": 2, "y": 4 },
+                { "matrix": [4, 3], "x": 3, "y": 4 }
```
Comment (fauxpark) [inline keyboards/hackpad/info.json:70]:
```suggestion
                {"matrix": [0, 3], "x": 3, "y": 0},

                {"matrix": [1, 0], "x": 0, "y": 1},
                {"matrix": [1, 1], "x": 1, "y": 1},
                {"matrix": [1, 2], "x": 2, "y": 1},
                {"matrix": [1, 3], "x": 3, "y": 1},

                {"matrix": [2, 0], "x": 0, "y": 2},
                {"matrix": [2, 1], "x": 1, "y": 2},
                {"matrix": [2, 2], "x": 2, "y": 2},
                {"matrix": [2, 3], "x": 3, "y": 2},

                {"matrix": [3, 0], "x": 0, "y": 3},
                {"matrix": [3, 1], "x": 1, "y": 3},
                {"matrix": [3, 2], "x": 2, "y": 3},
                {"matrix": [3, 3], "x": 3, "y": 3},

                {"matrix": [4, 0], "x": 0, "y": 4},
                {"matrix": [4, 1], "x": 1, "y": 4},
                {"matrix": [4, 2], "x": 2, "y": 4},
                {"matrix": [4, 3], "x": 3, "y": 4}
```

File: keyboards/hackpad/info.json [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    "diode_direction": "COL2ROW",
+    "features": {
+        "bootmagic": true,
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true,
+        "rgblight": true,
+        "encoder" : true
```
Comment (fauxpark) [inline keyboards/hackpad/info.json:15]:
```suggestion
        "encoder": true
```

File: keyboards/hackpad/info.json [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    },
+    "matrix_pins": {
+        "cols": ["B1", "B3", "B2", "B6"],
+        "rows": ["C6", "D7", "E6", "B4", "B5"]
+    },
+    "encoder" : {
+        "rotary" : [
+            {"pin_a" : "F5", "pin_b" : "F4"}
```
Comment (fauxpark) [inline keyboards/hackpad/info.json:23]:
```suggestion
    "encoder": {
        "rotary": [
            {"pin_a": "F5", "pin_b": "F4"}
```

File: keyboards/hackpad/info.json [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+            "rainbow_mood" : true,
+            "rainbow_swirl" : true,
+            "snake"   : true,
+            "knight"  : true,
+            "christmas" : true,
+            "static_gradient" : true,
+            "rgb_test" : true,
+            "alternating" : true,
+            "twinkle" : true
+        },
```
Comment (fauxpark) [inline keyboards/hackpad/info.json:49]:
```suggestion
    "rgblight": {
        "led_count": 16,
        "sleep": true,
        "animations": {
            "breathing": true,
            "rainbow_mood": true,
            "rainbow_swirl": true,
            "snake": true,
            "knight": true,
            "christmas": true,
            "static_gradient": true,
            "rgb_test": true,
            "alternating": true,
            "twinkle": true
        }
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/hackpad/readme.md (hunk 16 above)
Comment (ThePurox) [reply keyboards/hackpad/readme.md]: Fixed by the newest commit.
Thank you for noticing!

Comment (fauxpark) [reply keyboards/hackpad/readme.md]: Still doesn't work.
image

Comment (ThePurox) [reply keyboards/hackpad/readme.md]: Sorry for the annoyance.. Now the link points to the raw image, and the image is previewed.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/hackpad/readme.md [hunk 21]
```diff
@@ -0,0 +1,31 @@
+# hackpad
+
+![hackpad](https://git.imaginaerraum.de/Purox/Hackpad/raw/branch/main/assets/pic.jpg)
```
[the remark this answers is not in the document]
Comment (ThePurox) [reply keyboards/hackpad/readme.md:3]: Done

**Review by drashna — APPROVED** [review]

File: keyboards/hackpad/keymaps/default/keymap.c [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+     * │ 0 │ . │Ent│ / │
+     * └───┴───┴───┴───┘
+     */
+    [0] = LAYOUT(                                 LT(1, KC_MUTE),
+                            KC_7, KC_8,   KC_9,   DE_PLUS,
+                            KC_4, KC_5,   KC_6,   DE_MINS,
+                            KC_1, KC_2,   KC_3,   DE_ASTR,
+                            KC_0, KC_DOT, KC_ENT, DE_SLSH),
+
+    [1] = LAYOUT(                                         XXXXXXX,
```
Comment (waffle87) [inline keyboards/hackpad/keymaps/default/keymap.c:26]:
```suggestion
    [1] = LAYOUT(                                         _______,
```

**Review by waffle87 — APPROVED** [review]

**Pull request merged**